scripts/buf.pl

```diff
@@ -40,64 +42,60 @@
     for my $window (windows) {
-	next unless defined $window;
-	next if $window->{name} eq 'status';
-	my $view = $window->view;
-	my $line = $view->get_lines;
-	my $lines  = 0;
-	my $buf = '';
-	if (defined $line){
-	    {
-		$buf .= $line->get_text(1) . "\n";
-		$line = $line->next;
-		$lines++;
-		redo if defined $line;
-	    }
-	}
-	printf $fh "%s:%s\n%s", $window->{refnum}, $lines, $buf;
+        next unless defined $window;
+        next if $window->{name} eq 'status';
+        my $view = $window->view;
+        my $line = $view->get_lines;
+        my $lines  = 0;
+        my $buf = '';
+        my $output;
+        if (defined $line) {
+            {
+              push @$output, { level => $line->{info}{level}, data => $line->get_text(1) };
+              $line = $line->next;
+              redo if defined $line;
+            }
+        }
+        push @{$out->{windows}}, { refnum => $window->{refnum}, lines => $output };
     }
-    close $fh;
+    my $old_umask = umask 0077;
+    my $fn = _filename;
+    store($out, $fn) or die "Could not store data to $fn";
+    umask $old_umask;
     unlink sprintf("%s/sessionconfig", get_irssi_dir);
     command 'layout save';
     command 'save';
 }
 
 sub restore {
     my $fn = _filename;
-    open my $fh, q{<}, $fn or die "open $fn: $!";
+    my $in = retrieve($fn) or die "Could not retrieve data from $fn";
     unlink $fn or warn "unlink $fn: $!";
-
-    my @suppress = split /\0/, readline $fh;
-    if (settings_get_bool 'upgrade_suppress_join') {
-	chomp $suppress[-1];
-	@suppress{@suppress} = (2) x @suppress;
-    }
+  
+    my @suppress = @{$in->{suppress}};
+    @suppress{@suppress} = (2) x @suppress if (settings_get_bool 'upgrade_suppress_join');
+  
     active_win->command('^window scroll off');
-    while (my $bla = readline $fh){
-	chomp $bla;
-	my ($refnum, $lines) = split /:/, $bla;
-	next unless $lines;
-	my $window = window_find_refnum $refnum;
-	unless (defined $window){
-	    readline $fh for 1..$lines;
-	    next;
-	}
-	my $view = $window->view;
-	$view->remove_all_lines();
-	$view->redraw();
-	my $buf = '';
-	$buf .= readline $fh for 1..$lines;
-	my $sep = settings_get_str 'upgrade_separator';
-	$sep .= "\n" if $sep ne '';
-	$window->gui_printtext_after(undef, MSGLEVEL_CLIENTNOTICE, "$buf\cO$sep");
-	$view->redraw();
+    for my $win (@{$in->{windows}}) {
+        my $window = window_find_refnum $win->{refnum};
+        next unless $window;
+        my @lines  = @{ $win->{lines} || [] };
+        next unless @lines;
+    
+        my $view = $window->view;
+        $view->remove_all_lines();
+        $view->redraw();
+        for my $line (@lines) {
+            my $level = $line->{level};
+            my $data  = $line->{data};
+            $window->gui_printtext_after($window->last_line_insert, $level, "$data\n");
+        }
+        my $sep = settings_get_str 'upgrade_separator';
+        $sep .= "\n" if $sep ne '';
```

Comment:
this line seems to cause an empty line to appear on screen

Reply:
maybe just remove it?

Reply:
Maybe we should just remove it?

Reply:
Or remove the \n from 
$window->gui_printtext_after($window->last_line_insert, MSGLEVEL_CLIENTNOTICE, "\cO$sep\n");

Reply:
that seems buggy (bug was in the original script) if we have an empty $sep then we should not print anything (also not a \c0) otherwise we have an incomplete line hanging around in the buffer

Reply:
Alright, that should be it :)